src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -2575,19 +2568,29 @@ internal static void WriteConsole(ConsoleHandle consoleHandle, ReadOnlySpan<char
             }
 
             outBuffer = output.Slice(cursor);
+            if (!newLine)
+            {
+                WriteConsole(consoleHandle, outBuffer);
+                return;
+            }
 
-            if (newLine)
+            string lineEnding = Environment.NewLine;
+            int size = outBuffer.Length + lineEnding.Length;
+            if (size <= MaxStackAllocSize)
             {
-                var endOfLine = Environment.NewLine.AsSpan();
-                var endOfLineLength = endOfLine.Length;
-                Span<char> outBufferLine = stackalloc char[outBuffer.Length + endOfLineLength];
-                outBuffer.CopyTo(outBufferLine);
-                endOfLine.CopyTo(outBufferLine.Slice(outBufferLine.Length - endOfLineLength));
-                WriteConsole(consoleHandle, outBufferLine);
+                // We expect this code path to be hit more often, and thus optimize it with 'stackalloc'.
+                Span<char> buffer = stackalloc char[size];
+                outBuffer.CopyTo(buffer);
```

Comment:
Why do you want to create a constant size buffer when it may not be used at all? Only the `if (size <= MaxStackAllocSize)` code path will be using a stackalloc buffer, so it should be created here.

Reply:
It is hot path - it will be used almost always. If not, it's not that expensive. The benefit is that the compiler optimizes the code in this case - space is allocated before the method is called, otherwise special code will be added to lease the stack and release it.

Reply:
> Why do you want to create a constant size buffer when it may not be used at all? Only the `if (size <= MaxStackAllocSize)` code path will be using a stackalloc buffer, so it should be created here.

FWIW I have seen the runtime folks talk about this a bunch. Apparently, you should always allocate the max you're willing to and slice it down because dynamic stackalloc in particular is expensive. I have no sources though, just parroting runtime folks.

Reply:
> It is hot path - it will be used almost always. If not, it's not that expensive.

This is not true. When writing out the `prompt` string, it's always going through the `if (!newLine)` code path. And prompt string gets written out a lot in an interactive session.

> Apparently, you should always allocate the max you're willing to and slice it down because dynamic stackalloc in particular is expensive.

I compared the IL code generated from [stackalloc char[size]](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQOB2AhgLYCmAzgA4EDGJABAQEbU4DeOdndyAzN6nQAiJIgHsOXdti4zufZOgBs3ACx0ACgCcweYAAodwOmTAAvEgEoJsulJs2AylTwAeagAsCmgHx1GAVwAzQJJNOgBeYzN6F0j0VAAOOgB+Y2AaAGsCCAhRajoPLwBtE3MAXToQOhgSQIJ/CGAAbmt7TgUATj0A4NCAOgBJMgBRIgpgAE8LFul7AF9Whew5oA) vs. [stackalloc char[128]](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgAwFgBQO0EYccA7AQwFsBTAZwAdSBjSgAlICMGcBvHZv55AGYBqZgBFK5APa9+PbP0UDhyfADYBAFmYAFAE5hiwABSHgzamABelAJSylzeY8cBlesQA8DABak9AHzMbACuAGZhlHrMALwW1iyecfioABzMAPwWwIwA1qQQEFIMzL7+ANopqQC6zCDMMJRhpCEQwADcDi58oRFRscHhkXoAdK4QYEzG6FDxNradCt38qgCcxr3DIwCS1ACi5LTAAJ4LXYoAvl1X2BdAA=). The former generates 4 more IL instructions than the latter regarding the `stackalloc` operation, which is for multiplying 2 on `size` (the latter directly uses 256 and thus avoids the multiplication). However, the additional `buffer.Slice(0, size)` in the latter results in 5 more IL instructions. So, it's not clear to me the former is more expensive than the latter.

I think this perf difference may depend on the scenario context. I doubt it matters that much in this case.

Reply:
@SeeminglyScience Thanks for the code snippets.
I do see this pattern being used in a few places in dotnet code: https://grep.app/search?q=%5C%28stackalloc%20.%2A%5C%5B.%2A%5C%5D%5C%29%5C.Slice%5C%28&regexp=true

I will update the code to follow that pattern.

Reply:
@iSazonov and @SeeminglyScience, can you please share some pointers about why using `stackalloc` with a constant size is more performant than with a variable size?

Reply:
It is very difficult to find these comments in discussions :-(
That @SeeminglyScience mentioned https://github.com/dotnet/docs/issues/28823 - and I am ok to follow the pattern.
In the blog https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/#json it is mentioned [dotnet/runtime#55350](https://github.com/dotnet/runtime/pull/55350) 

I cannot find a comment about (tired1) optimization which move stack allocation before method call if stackalloc with constant is used.